### PR TITLE
fix(service-registry): deregister chains support updates CHAINS_PER_WORKER

### PR DESCRIPTION
**Pull request by maancham**
## Description

AXE-3920

## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: contracts/service-registry/src/contract/execute.rs [hunk 1]
```diff
@@ -156,9 +156,8 @@ pub fn deregister_chains_support(
         .may_load(deps.storage, (&service_name, &info.sender))?
         .ok_or(ContractError::WorkerNotFound)?;
 
-    for chain in chains {
-        WORKERS_PER_CHAIN.remove(deps.storage, (&service_name, &chain, &info.sender));
-    }
+    let _res =
+        state::deregister_chains_support(deps.storage, service_name.clone(), chains, info.sender);
```
Comment (cjcobb23) [inline contracts/service-registry/src/contract/execute.rs:160]:
```suggestion
        state::deregister_chains_support(deps.storage, service_name.clone(), chains, info.sender)?;
```
We shouldn't ignore results.

File: contracts/service-registry/src/state.rs [hunk 2]
```diff
@@ -184,6 +184,27 @@ pub fn register_chains_support(
...
+) -> Result<(), ContractError> {
+    CHAINS_PER_WORKER.update(storage, (&service_name, &worker), |current_chains| {
+        let mut current_chains = current_chains.unwrap_or_default();
+        for chain in chains.iter() {
+            current_chains.remove(chain);
+        }
+        Ok::<HashSet<ChainName>, ContractError>(current_chains)
+    })?;
+
+    for chain in chains.iter() {
```
Comment (cjcobb23) [inline contracts/service-registry/src/state.rs:201]:
```suggestion
    for chain in chains {
```

File: contracts/service-registry/src/state.rs [hunk 3]
```diff
@@ -184,6 +184,27 @@ pub fn register_chains_support(
...
+    service_name: String,
+    chains: Vec<ChainName>,
+    worker: WorkerAddress,
+) -> Result<(), ContractError> {
+    CHAINS_PER_WORKER.update(storage, (&service_name, &worker), |current_chains| {
+        let mut current_chains = current_chains.unwrap_or_default();
+        for chain in chains.iter() {
+            current_chains.remove(chain);
+        }
+        Ok::<HashSet<ChainName>, ContractError>(current_chains)
```
Comment (cjcobb23) [inline contracts/service-registry/src/state.rs:198]:
```suggestion
        Ok::<HashSet<ChainName>, ContractError>(
            current_chains
                .unwrap_or_default()
                .into_iter()
                .filter(|chain| !chains.contains(chain))
                .collect(),
        )
```

File: contracts/service-registry/src/state.rs [hunk 4]
```diff
@@ -184,6 +184,27 @@ pub fn register_chains_support(
     Ok(())
 }
 
+pub fn deregister_chains_support(
+    storage: &mut dyn Storage,
+    service_name: String,
+    chains: Vec<ChainName>,
+    worker: WorkerAddress,
+) -> Result<(), ContractError> {
+    CHAINS_PER_WORKER.update(storage, (&service_name, &worker), |current_chains| {
```
Comment (cjcobb23) [inline contracts/service-registry/src/state.rs:193]: This function allows workers to deregister for chains that they are not registered for. Is that intended? One side effect is that if a worker is not registered for any chains, this function is going to write an empty hashset into CHAINS_PER_WORKER

Comment (maancham) [reply contracts/service-registry/src/state.rs:193]: yeah it would not be an error.

**Pull request merged**